### ISSUE TITLE
Debugging guide and/or internal tracing for debugging purposes

### ISSUE THREAD

**Issue by davisjam** [libuv/libuv#1966]
While debugging libuv changes I often resort to printf's.

1. Is there a way to run a libuv `test/` within gdb and/or under valgrind? If so, can we put notes on how to do this in a `DEBUGGING.md`-type document?
2. Would there be interest in introducing some kind of internal tracing mechanism?

**Comment by cjihrig** [comment]
1. For valgrind, does something like `valgrind -v --tool=memcheck --leak-check=full --track-origins=yes --leak-resolution=high --show-reachable=yes --log-file=memcheck.log ./test/run-tests tmpdir tmpdir` work for you? I wouldn't be opposed to a wiki page or something for keeping that kind of information in one place.
2. Do you mean something like https://github.com/libuv/libuv/pull/1815?

**Comment by santigimeno** [comment]
> For valgrind, does something like valgrind -v --tool=memcheck --leak-check=full --track-origins=yes --leak-resolution=high --show-reachable=yes --log-file=memcheck.log ./test/run-tests tmpdir tmpdir work for you?

Haven't tried with valgrind but I'm not sure it will work as `./test/run-tests` is a script.

What works for me, at least wit gdb is:
```
$ ./gyp_uv.py -f make
$ make -C out
$ gdb --args out/Debug/run-tests test-name test-name
```

**Comment by davisjam** [comment]
> Would there be interest in introducing some kind of internal tracing mechanism?
>> Do you mean something like #1815?

OK, wrong word. I meant *logging* -- e.g. `UV__LOG(UV_LOG_FS, 4, "This is a log statement at verbosity level 4")`, with UV__LOG under a DEBUG ifdef.

I think (the splendid) #1815 is about monitoring the health of libuv from the embedder's perspective, not really what I had in mind.

**Comment by davisjam** [comment]
I ask about valgrind specifically because the tests have `MAKE_VALGRIND_HAPPY()` in them. So presumably someone knows how to do this!

**Comment by santigimeno** [comment]
>  I ask about valgrind specifically

I would use the command https://github.com/libuv/libuv/issues/1966#issuecomment-416939989 but using `out/Debug/run-tests` instead of `./test/run-tests`. For example, this works for me:

```
valgrind -v --tool=memcheck --leak-check=full --track-origins=yes --leak-resolution=high --show-reachable=yes --log-file=memcheck.log out/Debug/run-tests  udp_open udp_open
```

**Comment by cjihrig** [comment]
@santigimeno is right. FWIW, `test/.libs/run-tests` may also work.

**Comment by davisjam** [comment]
Thanks folks. See #1971.

FWIW, the command from @santigimeno works but only traces the execution of `run-tests`. For tests that involve a fork, we need to add `--trace-children=yes`.

For example, I deliberately inserted a memory leak into one of the libuv paths traversed by the `getaddrinfo_basic` test, and this invocation catches it:

```
valgrind -v --tool=memcheck --trace-children=yes --leak-check=full --track-origins=yes --leak-resolution=high --show-reachable=yes --log-file=memcheck.log out/Debug/run-tests getaddrinfo_basic
```

Without `--trace-children=yes`, no leak is reported.

**Comment by gireeshpunathil** [comment]
#1217 has used it in abundance (though used in conjunction with Node, practically it was calibrating libuv)

**Comment by davisjam** [comment]
Thanks @gireeshpunathil

**Comment by cjihrig** [comment]
Fixed by c0c672e1a0117b88a2b43a4472cfc50dcb33db45
